src/util/string.cpp

```diff
@@ -77,70 +76,65 @@ bool convert(const char *to, const char *from, char *outbuf,
 	}
 
 	iconv_close(cd);
+	*outbuf_size = old_outbuf_size - *outbuf_size;
 	return true;
 }
 
 #ifdef __ANDROID__
-// Android need manual caring to support the full character set possible with wchar_t
+// On Android iconv disagrees how big a wchar_t is for whatever reason
 const char *DEFAULT_ENCODING = "UTF-32LE";
 #else
 const char *DEFAULT_ENCODING = "WCHAR_T";
 #endif
 
 std::wstring utf8_to_wide(const std::string &input)
 {
-	size_t inbuf_size = input.length() + 1;
+	const size_t inbuf_size = input.length();
 	// maximum possible size, every character is sizeof(wchar_t) bytes
-	size_t outbuf_size = (input.length() + 1) * sizeof(wchar_t);
+	size_t outbuf_size = input.length() * sizeof(wchar_t);
 
-	char *inbuf = new char[inbuf_size];
+	char *inbuf = new char[inbuf_size]; // intentionally NOT null-terminated
 	memcpy(inbuf, input.c_str(), inbuf_size);
```

Comment:
Why making a copy?

Reply:
iconv only takes a mutable input buffer...

Reply:
It takes a `char **` but only reads buffer contents, so `const_cast<char *>` should be fine here. It does change the pointer itself though (thus couldn’t take `char const* const*`).

Reply:
Shouldn't it be taking a `const char **inbuf` then?

> but only reads buffer contents

I'm sure it does but the issue is that I don't see this guaranteed by the types or documentation.

Reply:
> Shouldn't it be taking a const char **inbuf then?

It wouldn’t be able to take `char **` then, there is no implicit conversion (even in C++; in C, rules are stricter so even `char **`→`char const* const*` requires a cast).